### ISSUE TITLE
Parameters don't copy between GPUs

### ISSUE THREAD

**Issue by YuJingBaoBao** [horovod/horovod#370]
I am trying to use horovod with tf slim ,  on 12 GPUs of 3 machines ( 4 GPUs each).
Then I use iperf3 to check the bandwidth, see nearly 0 bandwidth cost when training.

**Comment by YuJingBaoBao** [comment]
It seems parameters copy between GPUs on the same machine , but not between machines. Here are logs:
INFO:tensorflow:Recording summary at step 7363.
INFO:tensorflow:Recording summary at step 7363.
INFO:tensorflow:Recording summary at step 2433.
INFO:tensorflow:Recording summary at step 2433.
INFO:tensorflow:Recording summary at step 2433.
INFO:tensorflow:Recording summary at step 7363.
INFO:tensorflow:Recording summary at step 52730.
INFO:tensorflow:Recording summary at step 7363.
INFO:tensorflow:Recording summary at step 52730.

Here is my running command:
mpirun -np 12   -H ip1:4,ip2:4,ip3:4   -bind-to none -map-by slot     -x NCCL_DEBUG=DEBUG -x LD_LIBRARY_PATH -x PATH     -mca pml ob1 -mca btl ^openib python train_distribute_horovod.py 

I have set num_clones=1,num_replicas=1,num_ps_tasks=0,task=0 in my code.

**Comment by tgaddair** [comment]
Hey @YuJingBaoBao, do you see this same behavior when you try running the `tensorflow_mnist.py` examples in your environment?  When you print the rank and size for each process, do you get 12 distinct ranks and size 12?

```
print(hvd.rank(), hvd.size())
```

**Comment by YuJingBaoBao** [comment]
Thanks , @tgaddair  .
I have print hvd.rank() and hvd.size() , it's 12 distinct ranks and size 12.
And GPUs are working (GPU-Util nearly 100%)

**Comment by YuJingBaoBao** [comment]
I am not sure...

**Comment by YuJingBaoBao** [comment]
I have almost solved the problem by:
      bcast = hvd.broadcast_global_variables(0)
      update_ops.append(bcast)

Now processes don't save checkpoints.  I will continue to refuel. Thank you @tgaddair  @alsrgv

**Comment by dhkwon1122** [comment]
@YuJingBaoBao 
Hi, Did you solve the problem?
it seems like you modified the deeplab v3+ training code to use horovod,
can you share modified whole train.py?
Thank you in advance :D

**Comment by YuJingBaoBao** [comment]
add " optimizer=hvd.DistributedOptimizer(optimizer)   " before this line:
total_loss, grads_and_vars = model_deploy.optimize_clones( clones, optimizer) 

add 
"bcast = hvd.broadcast_global_variables(0)
update_ops.append(bcast)"
before this line:
"update_op = tf.group(*update_ops)"

the init_fn and init_op slim.learning.train are only called when start training from blank. So I add the broadcast in every step, not efficiency.

@dhkwon1122

**Comment by dhkwon1122** [comment]
@YuJingBaoBao
Sorry to bother you,
I modified the code as you mentioned but got messages like below,

WARNING: One or more tensors were submitted to be reduced, gathered or broadcasted by subset of ranks and are waiting for remainder of ranks for more than 60 seconds. This may indicate that different ranks are trying to submit different tensors or that only subset of ranks is submitting tensors, which will cause deadlock. Stalled ops: DistributedDistributedOptimizer_Allreduce/HorovodAllreduce_gradients_AddN_48_0 [ready ranks: 3, 2], DistributedDistributedOptimizer_Allreduce/HorovodAllreduce_gradients_MobilenetV2_Conv_BatchNorm_FusedBatchNorm_grad_tuple_control_dependency_2_0 [ready ranks: 3, 2], DistributedDistributedOptimizer_Allreduce/HorovodAllreduce_gradients_MobilenetV2_expanded_conv_depthwise_depthwise_grad_tuple_control_dependency_1_0 [ready ranks: 3, 2], DistributedDistributedOptimizer_Allreduce/HorovodAllreduce_gradients_MobilenetV2_expanded_conv_depthwise_BatchNorm_FusedBatchNorm_grad_tuple_control_dependency_1_0 [ready ranks: 3, 2], DistributedDistributedOptimizer_Allreduce/HorovodAllreduce_gradients_MobilenetV2_expanded_conv_project_BatchNorm_FusedBatchNorm_grad_tuple_control_dependency_1_0 [ready ranks: 3, 2], DistributedDistributedOptimizer_Allreduce/HorovodAllreduce_gradients_MobilenetV2_expanded_conv_1_expand_BatchNorm_FusedBatchNorm_grad_tuple_control_dependency_1_0 [ready ranks: 3, 2], DistributedDistributedOptimizer_Allre..........

So, I can not sure it is trained properly, T^T 
the 2 line is all you fixed?

**Comment by YuJingBaoBao** [comment]
add this before slim.learning.train:
    session_config.gpu_options.allow_growth = True
    session_config.gpu_options.visible_device_list=str(hvd.local_rank())

@dhkwon1122

**Comment by dhkwon1122** [comment]
@YuJingBaoBao 

Totally, my modification is below
1. import horovod.tensorflow and do hvd.init()
2. add " optimizer=hvd.DistributedOptimizer(optimizer) " before this line:
total_loss, grads_and_vars = model_deploy.optimize_clones( clones, optimizer)
3. add
"bcast = hvd.broadcast_global_variables(0)
update_ops.append(bcast)"
before this line:
"update_op = tf.group(*update_ops)"
4. add this before slim.learning.train:
session_config.gpu_options.allow_growth = True
session_config.gpu_options.visible_device_list=str(hvd.local_rank())

And I can't see the log uploaded by you.
"""
INFO:tensorflow:loss = 5.972283e-07, step = 2642, rank = 6.0 (2.084 sec)
INFO:tensorflow:loss = 0.001997095, step = 2642, rank = 7.0 (2.087 sec)
INFO:tensorflow:loss = 0.14977533, step = 140, rank = 3.0 (2.090 sec)
INFO:tensorflow:loss = 0.37332544, step = 140, rank = 1.0 (2.084 sec)
INFO:tensorflow:loss = 0.18596107, step = 2642, rank = 5.0 (2.084 sec)
INFO:tensorflow:loss = 0.3339143, step = 140, rank = 0.0 (2.084 sec)
INFO:tensorflow:loss = 0.023410445, step = 2642, rank = 4.0 (2.084 sec)
"""

just got really complicated broadcast message ...
"""
WARNING: One or more tensors were submitted to be reduced, gathered or broadcasted by subset of ranks and are waiting for remainder of ranks for more than 60 seconds. This may indicate that different ranks are trying to submit different tensors or that only subset of ranks is submitting tensors, which will cause deadlock. Stalled ops: DistributedDistributedOptimizer_Allreduce/HorovodAllreduce_gradients_AddN_48_0 [ready ranks: 3, 2], 
.....
"""

How can I sure it is trained...? I m so wondering how to gather gradients and update it.... T^T

**Comment by YuJingBaoBao** [comment]
@dhkwon1122 
I am sorry. I don't know why.

**Comment by YuJingBaoBao** [comment]
When I run tensorflow_mnist.py, There is INFO like this at the begining:
INFO Ring 00 : 0 1 2 3 4 5 6 7 8 9 10 11
But when I run train_distribute_horovod.py, there is not. Is that normal?  @tgaddair @alsrgv

**Comment by alsrgv** [comment]
@YuJingBaoBao, @dhkwon1122, I would recommend checking #220 for the better way of broadcasting variables with TF-Slim.
